### DISCUSSION TITLE
Beginner: Help with changing player input and best practice for changing player state

### DISCUSSION THREAD

**Discussion by ckamarga** [nathanhoad/godot_dialogue_manager#425] [Q&A] [answered]
Hello,

I downloaded this dialogue manager yesterday (which is amazing so far so kudos nathanhoad!) and I've been trying to smoothly integrate it into my project. I've managed to get an NPC of mine to trigger the dialogue and used mutations to help change the player's state (e.g. to and from dialogue state).

But I've kinda hit a bit of a head scratcher. These may be simple questions but I'm still somewhat of a beginner with Godot and I'm more a designer than programmer so you may have to dumb things down for me.

Anyway, I'm trying to do things one step at a time and here is currently what I'm trying to achieve:

1. I figured out how to change the input that skips the text typing (by changing the string for skip_action), but can't figure out how to change the input that let's me load the next dialogue—effectively, I want to transfer all mouse inputs into a controller input
2. When my player interacts with an NPC, I have it trigger a function that calls DialogueManager.show_dialogue_balloon() and changes the player's state to DIALOGUE (so they can't do anything during dialogue) and then inside the dialogue script, before '=> END', I set the player's state back to its original—is this the best way of doing this? Because I'll have options where some can end the dialogue and I don't want to have to keep copy pasting that line of code for each '=> END' I have in the dialogue script

Thanks in advance!

p.s.
Speaking of skip_action, what does the '&' mean/do at the start of the string value?

**Comment by nathanhoad** [comment, marked as answer]
1. The example balloon has full keyboard and controller support. The inputs are mapped to "ui_accept" and "ui_cancel" by default. You can change the actual keys/buttons for these actions in the Input Map in Project Settings.
2. You can connect to the `DialogueManager.dialogue_ended` signal to be notified when the dialogue has finished.
3. The `&` in front of a string means it's a StringName (a special kind of string that always occupies the same place in memory).

**Reply by ckamarga** [reply]
Thanks nathanhoad! I really appreciate your reply.

I ended up changing the input in the code rather than the Input Map under Project Settings: I didn't realize that the bubble is an instantiated scene so I found the code that controls what button continues the dialogue. Hopefully this is fine.

Anyway, thanks again for creating this amazing dialogue system. It's been a HUGE help to my project. I've made sure to include you in the credits (assuming I ever finish and publish my game).

Happy New Year!
